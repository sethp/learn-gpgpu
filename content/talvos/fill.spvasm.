; SPIR-V
; Version: 1.3
OpCapability Shader
OpMemoryModel Logical GLSL450
OpEntryPoint GLCompute %fill_fn "FILL" %gl_GlobalInvocationID

  ; metadata (& pseudo-linker)
OpDecorate %gl_GlobalInvocationID BuiltIn GlobalInvocationId
OpDecorate %_arr_uint32_t ArrayStride 4
OpDecorate %buf DescriptorSet 0
OpDecorate %buf Binding 0

  ; types
%uint32_t = OpTypeInt 32 0
  %void_t = OpTypeVoid
%gbl_id_t = OpTypeVector %uint32_t 3

%_arr_uint32_t = OpTypeRuntimeArray %uint32_t
		%void_fn_t = OpTypeFunction %void_t

%_ptr_StorageBuffer_uint32_t = OpTypePointer StorageBuffer %uint32_t
%_arr_StorageBuffer_uint32_t = OpTypePointer StorageBuffer %_arr_uint32_t
        %_ptr_Input_gbl_id_t = OpTypePointer Input %gbl_id_t
        %_ptr_Input_uint32_t = OpTypePointer Input %uint32_t


  ; global arguments & constants
              %const_0 = OpConstant %uint32_t 0
                 %FILL = OpConstant %uint32_t 1
%gl_GlobalInvocationID = OpVariable %_ptr_Input_gbl_id_t Input
                  %buf = OpVariable %_arr_StorageBuffer_uint32_t StorageBuffer

  ; FILL entry point
 %fill_fn = OpFunction %void_t None %void_fn_t
       %1 = OpLabel
       %2 = OpAccessChain %_ptr_Input_uint32_t %gl_GlobalInvocationID %const_0
       %3 = OpLoad %uint32_t %2
       %4 = OpAccessChain %_ptr_StorageBuffer_uint32_t %buf %3
            OpStore %4 %FILL
            OpReturn
            OpFunctionEnd
