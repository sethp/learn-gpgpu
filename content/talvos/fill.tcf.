# MODULE fill.spvasm
# ENTRY fill

BUFFER a 64 UNINIT
#         ^ bytes

DESCRIPTOR_SET 0 0 0 a

DISPATCH 16 1 1
#         ^ work items (here, "elements")

DUMP UINT32 a
#       ^ bits

# NB: a's size and the DISPATCH must agree;
#    i.e. 16 elements == (64 bytes / (32 bits per element / 8 bits per byte))
# a should be filled with the value of %FILL
