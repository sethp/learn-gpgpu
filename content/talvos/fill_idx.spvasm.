; SPIR-V
; Version: 1.5
OpCapability Kernel
OpCapability BuffersTALVOS
OpCapability ExecTALVOS
OpCapability PhysicalStorageBufferAddresses
OpExtension "SPV_TALVOS_buffers"
OpExtension "SPV_TALVOS_exec"
OpMemoryModel Logical OpenCL

OpEntryPoint Kernel %main_fn "main" %gl_GlobalInvocationID

OpExecutionGlobalSizeTALVOS %main_fn 16 1 1

OpDecorate %gl_GlobalInvocationID BuiltIn GlobalInvocationId
OpDecorate %_arr_uint32_t ArrayStride 4

; types
                              %void_t = OpTypeVoid
                           %void_fn_t = OpTypeFunction %void_t
                            %uint32_t = OpTypeInt 32 0
                            %gbl_id_t = OpTypeVector %uint32_t 3

                            %arr_len = OpConstant %uint32_t 16
                      %_arr_uint32_t = OpTypeArray %uint32_t %arr_len

                %_ptr_Input_gbl_id_t = OpTypePointer Input %gbl_id_t
                %_ptr_Input_uint32_t = OpTypePointer Input %uint32_t

%_ptr_PhysicalStorageBuffer_uint32_t = OpTypePointer PhysicalStorageBuffer %uint32_t
%_arr_PhysicalStorageBuffer_uint32_t = OpTypePointer PhysicalStorageBuffer %_arr_uint32_t


; global arguments & constants
                    %n = OpConstant %uint32_t 0
%gl_GlobalInvocationID = OpVariable %_ptr_Input_gbl_id_t Input
                 %buf0 = OpBufferTALVOS %_arr_PhysicalStorageBuffer_uint32_t PhysicalStorageBuffer 64 "a"

; FILL_IDX entry point
%main_fn = OpFunction %void_t None %void_fn_t
      %1 = OpLabel
      %2 = OpAccessChain %_ptr_Input_uint32_t %gl_GlobalInvocationID %n
      %3 = OpLoad %uint32_t %2 Aligned 4
      %4 = OpAccessChain %_ptr_PhysicalStorageBuffer_uint32_t %buf0 %3
           OpStore %4 %3 Aligned 4
           OpReturn
           OpFunctionEnd
