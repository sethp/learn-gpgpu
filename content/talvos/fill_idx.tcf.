# MODULE fill_idx.spvasm
# ENTRY main

EXEC

