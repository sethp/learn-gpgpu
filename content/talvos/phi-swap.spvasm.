OpCapability Shader
OpCapability VariablePointers
OpExtension "SPV_KHR_storage_buffer_storage_class"
OpExtension "SPV_KHR_variable_pointers"
OpMemoryModel Logical GLSL450
OpEntryPoint GLCompute %1 "phi-swap" %3
OpExecutionMode %1 LocalSize 1 1 1
%2 = OpString "filename"
OpDecorate %3 DescriptorSet 0
OpDecorate %3 Binding 0
OpDecorate %4 Block
OpDecorate %9 ArrayStride 4
OpMemberDecorate %4 0 Offset 0

%5 = OpTypeVoid
%6 = OpTypeFunction %5
%7 = OpTypeBool
%8 = OpTypeInt 32 0
%9 = OpTypeRuntimeArray %8
%4 = OpTypeStruct %9
%10 = OpTypePointer StorageBuffer %4
%11 = OpTypePointer StorageBuffer %8

%3 = OpVariable %10 StorageBuffer

%12 = OpConstantTrue %7
%13 = OpConstantFalse %7
%14 = OpConstant %8 42
%15 = OpConstant %8 7
%16 = OpConstant %8 0
%17 = OpConstant %8 1


%1 = OpFunction %5 None %6

%18 = OpLabel
OpBranch %19
%19 = OpLabel
%20 = OpPhi %7 %12 %18 %13 %19
%21 = OpPhi %8 %14 %18 %22 %19

OpLine %2 0 0
%22 = OpPhi %8 %15 %18 %21 %19
OpLoopMerge %23 %19 None
OpBranchConditional %20 %19 %23

%23 = OpLabel
%24 = OpAccessChain %11 %3 %16 %16
OpStore %24 %21
%25 = OpAccessChain %11 %3 %16 %17
OpStore %25 %22
OpReturn

OpFunctionEnd
