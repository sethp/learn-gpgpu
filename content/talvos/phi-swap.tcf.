ENTRY phi-swap

BUFFER output 8 FILL INT32 0

DESCRIPTOR_SET 0 0 0 output

DISPATCH 1 1 1

DUMP INT32 output

# CHECK: Buffer 'output' (8 bytes):
# CHECK: output[0] = 7
# CHECK: output[1] = 42
