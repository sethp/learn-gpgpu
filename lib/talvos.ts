import { eagerDeref } from "./binding";
import { Ptr } from "./binding/ptr";
import { UTF8ArrayToString, stringToUTF8Array } from "./binding/strings";
import { std$$deque, std$$optional, std$$string, std$$vector } from "./stlBinding";

const LITTLE_ENDIAN = true; // yay booleans

const NULLPTR = 0;

export class Talvos$$Params {
	constructor(public ptr: Ptr) { }
	static get SIZE() {
		return 64;
	}

	get EntryName() {
		return UTF8ArrayToString(this.ptr.slice(0, 64).as(Uint8Array), 0, 64);
	}
	set EntryName(name: string) {
		const written = stringToUTF8Array(name, this.ptr.slice(0, 64).as(Uint8Array), 0, 64);
		if (written < name.length) {
			throw new Error('out of memory: string too long');
		}
	}
}

export class Talvos$$EntryPoint {
	constructor(public ptr: Ptr) { }
	static get SIZE() {
		return 36;
	}

	get Name() {
		return new std$$string(this.ptr.slice(4, 4 + std$$string.SIZE));
	}
}

// std::vector<talvos::EntryPoint *>
// --> Iterable<Talvos$$EntryPoint>
export class _EntryPoints extends std$$vector(eagerDeref(Talvos$$EntryPoint)) {
}

export class Talvos$$Buffer {
	constructor(public ptr: Ptr) { }
	static get SIZE() {
		return 32;
	}

	get Id() {
		return this.ptr.data.getUint32(0, LITTLE_ENDIAN);
	}

	get Size() {
		return this.ptr.data.getUint32(8, LITTLE_ENDIAN);
	}

	get Name() {
		const rTy = std$$optional(std$$string);
		return new (rTy)(this.ptr.slice(16, 16 + rTy.SIZE));
	}
}

export class Talvos$$Module {
	constructor(public ptr: Ptr) { }
	static get SIZE() {
		return 128;
	}

	get Buffers() {
		return new (std$$vector(Talvos$$Buffer))(this.ptr.slice(116, 116 + std$$vector.SIZE));
	}

	get EntryPoints() {
		return new _EntryPoints(this.ptr.slice(40, 40 + _EntryPoints.SIZE));
	}
}

class Talvos$$Memory$$Alloc {
	constructor(public ptr: Ptr) { }
	static get SIZE() {
		return 16;
	}

	get NumBytes(): bigint {
		return this.ptr.data.getBigUint64(0, LITTLE_ENDIAN);
	}

	get Data(): Ptr {
		return this.ptr.deref(8, Number(this.NumBytes), LITTLE_ENDIAN)
	}
}

export class Talvos$$Memory {
	constructor(public ptr: Ptr) { }
	static get SIZE() {
		return 2456;
	}

	// TODO[seth]: it would be nice, at some point, to think about how to bind this to the C++ impl
	// (unfortunately, that means making the whole enchilada async, don't it?)
	// is there any way to late-bind the method impls?

	// TODO[seth]: wrap device-side addresses in a type? and/or distinguish device-side `Ptr`s somehow?
	// uh oh, number isn't precise enough lololol; I guess that solves that (kinda)

	// cf. `Memory::load` (which `memcpy`s the data)
	deref(Addr: bigint, Size: number): Ptr {
		const BUFFER_BITS = 16n;
		const OFFSET_BITS = 48n;

		const Id = Addr >> OFFSET_BITS;
		const Offset = Number(Addr & BigInt.asUintN(64, (2n ** 64n - 1n) >> BUFFER_BITS));

		const Alloc = this.Allocs.get(Number(Id));
		console.assert(Offset + Size <= Alloc.NumBytes)

		return Alloc.Data.slice(Offset, Size);
	}

	private get Allocs() {
		return new (std$$vector(Talvos$$Memory$$Alloc))(this.ptr.slice(2432, 2432 + std$$vector.SIZE))
	}

}

export class Talvos$$Object {
	constructor(public ptr: Ptr) { }
	static get SIZE() {
		return 20;
	}

	get Data(): Ptr {
		return this.ptr.deref(4, undefined /* TODO the type knows */, LITTLE_ENDIAN)
	}
}

export class Talvos$$PipelineExecutor {
	constructor(public ptr: Ptr) { }
	static get SIZE() {
		return 272;
	}

	get Objects() {
		return new (std$$vector(Talvos$$Object))(this.ptr.slice(32, 32 + std$$vector.SIZE))
	}
}

export class Talvos$$Device {
	constructor(public ptr: Ptr) { }
	static get SIZE() {
		return 112;
	}

	get GlobalMemory() {
		return new Talvos$$Memory(this.ptr.deref(16, Talvos$$Memory.SIZE, LITTLE_ENDIAN));
	}

	get PipelineExecutor() {
		return new Talvos$$PipelineExecutor(this.ptr.deref(32, Talvos$$PipelineExecutor.SIZE, LITTLE_ENDIAN));
	}
}

export class Talvos$$Dim3 {
	constructor(public ptr: Ptr) { }
	static get SIZE() {
		return 12;
	}

	get X() { return this.ptr.data.getUint32(0, LITTLE_ENDIAN); }
	get Y() { return this.ptr.data.getUint32(4, LITTLE_ENDIAN); }
	get Z() { return this.ptr.data.getUint32(8, LITTLE_ENDIAN); }

	// fixme: this could be a CArray type or something in lib/binding
	get(idx: number) {
		const elemSize = 4;
		const n = 3;
		if (idx < 0 || idx > n) {
			throw new Error(`out of bounds access: for index ${idx} with elements [0..${n})`);
		}
		const offset = idx * elemSize;
		return this.ptr.data.getUint32(offset, LITTLE_ENDIAN);
	}

	[Symbol.iterator]() {
		const [start, end] = this.ptr.reslice(Talvos$$Dim3.SIZE);
		const elemSize = 4;
		var itr = start;
		return {
			next() {
				if (itr.addr >= end.addr) return { done: true, value: undefined as never };

				const ret = {
					done: false,
					value: itr.data.getUint32(0, LITTLE_ENDIAN),
				};
				itr = itr.slice(elemSize);
				return ret;
			},
		};
	}

}

export class Talvos$$Instruction {
	constructor(public ptr: Ptr) { }
	static get SIZE() {
		return 20;
	}

	// hmm, nulls.
	// TODO: what's at the very beginning of memory? can we ban pointers to it at construction?
	hasResultType() { return this.ptr.getUsize(0, LITTLE_ENDIAN) != NULLPTR; }
	get ResultType() {
		return new Talvos$$Type(this.ptr.deref(0, Talvos$$Type.SIZE, LITTLE_ENDIAN));
	}

	get Opcode() {
		return this.ptr.data.getUint16(4, LITTLE_ENDIAN);
	}

	get NumOperands() {
		return this.ptr.data.getUint16(6, LITTLE_ENDIAN);
	}

	get Operands() {
		// fixme: this could be a SizedCArray type or something in lib/binding
		const elemSize = 4;
		const n = this.NumOperands;
		const ptr = this.ptr.deref(8, elemSize * n, LITTLE_ENDIAN);
		return new class {
			get(idx: number) {
				if (idx < 0 || idx > n) {
					throw new Error(`out of bounds access: for index ${idx} with elements [0..${n})`);
				}
				const offset = idx * elemSize;
				return ptr.data.getUint32(offset, LITTLE_ENDIAN);
			}

			[Symbol.iterator]() {
				const [start, end] = ptr.reslice(ptr.byteLength);
				var itr = start;
				return {
					next() {
						if (itr.addr >= end.addr) return { done: true, value: undefined as never };

						const ret = {
							done: false,
							value: itr.data.getUint32(0, LITTLE_ENDIAN),
						};
						itr = itr.slice(elemSize);
						return ret;
					},
				};
			}
		}
	}
}

export class Talvos$$Type {
	constructor(public ptr: Ptr) { }
	static get SIZE() {
		return 88;
	}
}

export class Talvos$$Core {
	constructor(public ptr: Ptr) { }
	static get SIZE() {
		return 28;
	}

	get Microtasks() {
		return new (std$$deque(undefined as any))(this.ptr.slice(4, 4 + std$$deque.SIZE))
	}

	get PC() {
		// nullable
		let tgt = this.ptr.deref(0, Talvos$$Instruction.SIZE, LITTLE_ENDIAN);
		return tgt.addr == 0 ? null : new Talvos$$Instruction(tgt);
	}
}

export class Talvos$$Cores extends std$$vector(Talvos$$Core) { }
