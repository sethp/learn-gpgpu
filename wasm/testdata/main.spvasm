; SPIR-V
; Version: 1.3
OpCapability Shader  ; TODO we need this for descriptors? really?
OpCapability Kernel
OpCapability DispatchTALVOS
OpExtension "SPV_TALVOS_dispatch"
OpMemoryModel Logical OpenCL
OpEntryPoint Kernel %main_fn "main" %gl_GlobalInvocationID
OpEntryPoint Kernel %fill_fn "FILL" %gl_GlobalInvocationID
OpEntryPoint Kernel %sers_fn "SERIES" %gl_GlobalInvocationID

  ; metadata (& pseudo-linker)
OpDecorate %gl_GlobalInvocationID BuiltIn GlobalInvocationId
; OpDecorate %_arr_uint32_t ArrayStride 4
OpDecorate %buf0 DescriptorSet 0
OpDecorate %buf0 Binding 0
OpDecorate %buf1 DescriptorSet 0
OpDecorate %buf1 Binding 1

  ; types
   %void_t = OpTypeVoid
%void_fn_t = OpTypeFunction %void_t
 %uint32_t = OpTypeInt 32 0
 %gbl_id_t = OpTypeVector %uint32_t 3

%_arr_uint32_t = OpTypeRuntimeArray %uint32_t

%_ptr_StorageBuffer_uint32_t = OpTypePointer StorageBuffer %uint32_t
%_arr_StorageBuffer_uint32_t = OpTypePointer StorageBuffer %_arr_uint32_t
        %_ptr_Input_gbl_id_t = OpTypePointer Input %gbl_id_t
        %_ptr_Input_uint32_t = OpTypePointer Input %uint32_t


  ; global arguments & constants
              %const_0 = OpConstant %uint32_t 0
                 %FILL = OpConstant %uint32_t 1
%gl_GlobalInvocationID = OpVariable %_ptr_Input_gbl_id_t Input
                 %buf0 = OpVariable %_arr_StorageBuffer_uint32_t StorageBuffer
                 %buf1 = OpVariable %_arr_StorageBuffer_uint32_t StorageBuffer

  ; main entry point
 %main_fn = OpFunction %void_t None %void_fn_t
       %0 = OpLabel
			 			; OpFillTALVOS <param> ; fills with <param>s
			 			OpDispatchTALVOS ;  %fill_fn 16 1 1
						OpDispatchTALVOS ;  %sers_fn 16 1 1
            OpReturn
            OpFunctionEnd

  ; FILL entry point
 %fill_fn = OpFunction %void_t None %void_fn_t
      %f1 = OpLabel
      %f2 = OpAccessChain %_ptr_Input_uint32_t %gl_GlobalInvocationID %const_0
      %f3 = OpLoad %uint32_t %f2
      %f4 = OpAccessChain %_ptr_StorageBuffer_uint32_t %buf0 %f3
            OpStore %f4 %FILL
            OpReturn
            OpFunctionEnd

  ; SERIES entry point
 %sers_fn = OpFunction %void_t None %void_fn_t
      %s1 = OpLabel
      %s2 = OpAccessChain %_ptr_Input_uint32_t %gl_GlobalInvocationID %const_0
      %s3 = OpLoad %uint32_t %s2
      %s4 = OpAccessChain %_ptr_StorageBuffer_uint32_t %buf1 %s3
            OpStore %s4 %s3
            OpReturn
            OpFunctionEnd
