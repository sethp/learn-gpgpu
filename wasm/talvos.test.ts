import { describe, expect, test } from 'vitest';
import talvos from './talvos-wasm';
import main from './testdata/main.spvasm?raw'
// TODO: replace with main.spvasm ?
import fill from '../content/talvos/fill.spvasm.?raw'

describe('talvos', () => {
	test('inits', async () => {
		await talvos();
	})

	test('exception', async () => {
		let instance = await talvos();

		expect(() => instance.ccall('exception'))
			.toThrowErrorMatchingSnapshot();
	})
	test('assertion', async () => {
		let [stdout, stderr] = ['', ''];
		let instance = await talvos({
			print: (text: any) => { stdout += text; },
			printErr: (text: any) => { stderr += text; },
		});

		expect(() => instance.cwrap('assertion')())
			.toThrowErrorMatchingSnapshot();

		expect(stdout).toBe('');
		// TODO: this is sensitive to the line number of the assert
		expect(stderr).toMatchSnapshot();
	})
	test('oob_write', async () => {
		let [stdout, stderr] = ['', ''];
		let instance = await talvos({
			print: (text: any) => { stdout += text + "\n"; },
			printErr: (text: any) => { stderr += text + "\n"; },
		});


		instance.cwrap('test_entry', 'void', ['string', 'string', 'string'])(
			fill,
			'FILL',
			`
BUFFER a 64 UNINIT
DESCRIPTOR_SET 0 0 0 a

DISPATCH 17 1 1
			`
		)

		expect(stdout).to.be.empty;
		expect(stderr).toMatchInlineSnapshot(`
			"
			Invalid store of 4 bytes to address 0x1000000000040 (Device scope)
			    Entry point: %1 FILL
			    Invocation: Global(16,0,0) Local(0,0,0) Group(16,0,0)
			      OpStore %18 %14

			"
		`);
	})
	test('main kernel', async () => {
		let [stdout, stderr] = ['', ''];
		let instance = await talvos({
			print: (text: any) => { stdout += text + '\n'; },
			printErr: (text: any) => { stderr += text; console.error(text) },
		});

		instance.cwrap('test_entry', 'void', ['string', 'string', 'string'])(
			main,
			`main`,
			`
			BUFFER fill 64 UNINIT
			BUFFER series 64 UNINIT

			DESCRIPTOR_SET 0 0 0 fill
			DESCRIPTOR_SET 0 1 0 series

			#DISPATCH 16 1 1
			DISPATCH 1 1 1

			DUMP UINT32 fill
			DUMP UINT32 series
			`
		)

		expect(stderr).toEqual('');
		expect(stdout).toMatchInlineSnapshot(`
			"
			Buffer 'fill' (64 bytes):
			  fill[0] = 431188
			  fill[1] = 431188
			  fill[2] = 8
			  fill[3] = 8
			  fill[4] = 24
			  fill[5] = 26
			  fill[6] = 508392
			  fill[7] = 10
			  fill[8] = 10
			  fill[9] = 0
			  fill[10] = 48
			  fill[11] = 26
			  fill[12] = 512576
			  fill[13] = 8
			  fill[14] = 8
			  fill[15] = 0

			Buffer 'series' (64 bytes):
			  series[0] = 431124
			  series[1] = 73
			  series[2] = 512944
			  series[3] = 431164
			  series[4] = 33
			  series[5] = 1
			  series[6] = 144
			  series[7] = 49
			  series[8] = 512544
			  series[9] = 508608
			  series[10] = 6
			  series[11] = 6
			  series[12] = 24
			  series[13] = 26
			  series[14] = 0
			  series[15] = 6
			"
		`);
	})
})
