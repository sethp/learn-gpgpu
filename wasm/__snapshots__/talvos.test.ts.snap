// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`talvos > assertion 1`] = `[RuntimeError: unreachable]`;

exports[`talvos > assertion 2`] = `"Aborted(Assertion failed: false && "hello: it's an assertion!", at: ../wasm/talvos/tools/talvos-cmd/wasm.cpp,472,assertion)"`;

exports[`talvos > exception 1`] = `
Exception {
  "message": [
    "std::runtime_error",
    "hello: it's an exception!",
  ],
}
`;
